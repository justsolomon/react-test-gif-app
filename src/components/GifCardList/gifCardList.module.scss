@import "../../assets/scss/base/variables";

.card-list {
  display: flex;
  margin-left: -1rem;
  width: auto;

  &__column {
    padding-left: 1rem;
    background-clip: padding-box;

    & > div {
      margin-bottom: 1rem;
    }
  }

  @media #{$media-mobile-sm} {
    margin-left: -0.5rem;

    &__column {
      padding-left: 0.5rem;

      & > div {
        margin-bottom: 0.5rem;
      }
    }
  }
}

.search-header {
  font-size: $font-size--xxl;
  margin-bottom: 1rem;

  span {
    font-weight: bold;
  }

  @media #{$media-mobile} {
    font-size: $font-size--xl;
    line-height: $font-size--xl;
  }

  @media #{$media-mobile-sm} {
    font-size: $font-size--lg;
  }
}
