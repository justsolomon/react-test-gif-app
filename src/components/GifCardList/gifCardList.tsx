import GifCard from "components/GifCard/gifCard";
import { IGif } from "types/global";
import styles from "./gifCardList.module.scss";
import Masonry from "react-masonry-css";
import Spinner from "components/global/Spinner";

interface IGifCardList {
  gifs: IGif[];
  loaderRef: (node?: Element | null) => void;
  nextPage: boolean;
}

const GifCardList = ({
  gifs,
  loaderRef,
  nextPage
}: IGifCardList): JSX.Element => {
  const breakpoints = {
    default: 5,
    1100: 4,
    768: 3,
    480: 2
  };

  return (
    <Masonry
      breakpointCols={breakpoints}
      className={styles["card-list"]}
      columnClassName={styles["card-list__column"]}
      data-testid="card-list"
    >
      {gifs.map((gif, index) => (
        <GifCard gif={gif} key={index} />
      ))}
      {nextPage ? (
        <div ref={loaderRef} data-testid="card-list__loader">
          <Spinner />
        </div>
      ) : null}
    </Masonry>
  );
};

export default GifCardList;
