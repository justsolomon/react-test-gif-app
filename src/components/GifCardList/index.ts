export { default as GifCardList } from "./gifCardList";
