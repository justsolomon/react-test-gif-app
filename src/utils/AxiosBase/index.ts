import axios from "axios";

// Creates a base instance for all axios based request
const instance = axios.create({
  baseURL: `https://api.giphy.com/v1/`
});

instance.interceptors.request.use((config) => {
  //add api key to query params
  config.params["api_key"] = process.env.REACT_APP_API_KEY;

  return config;
});

export default instance;
