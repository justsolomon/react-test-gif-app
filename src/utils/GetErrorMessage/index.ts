import { AxiosError } from "axios";

const getErrorMessage = (err: AxiosError): string => {
  let errorMessage;
  if (err.response) errorMessage = err.response.data.message;
  else errorMessage = err.message;

  return errorMessage;
};

export default getErrorMessage;
