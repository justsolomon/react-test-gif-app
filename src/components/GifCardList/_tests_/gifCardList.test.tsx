import { render, screen } from "@testing-library/react";
import { gif } from "utils/testUtils";
import GifCardList from "../gifCardList";

const renderGifCardList = (nextPage: boolean) => {
  return render(
    <GifCardList
      gifs={new Array(5).fill(gif)}
      nextPage={nextPage}
      loaderRef={() => null}
    />
  );
};

describe("<GifCardList />", () => {
  it("should render the GIF card list component properly", () => {
    const { asFragment } = renderGifCardList(false);

    expect(asFragment()).toMatchSnapshot();
  });
  it("should render the loader if next page exists", () => {
    renderGifCardList(true);

    const loader = screen.queryByTestId("card-list__loader");

    expect(loader).toBeInTheDocument();
    expect(loader?.querySelector("svg")).toBeInTheDocument();
  });
  it("should not render the loader if next page doesn't exist", () => {
    renderGifCardList(false);

    expect(screen.queryByTestId("card-list__loader")).not.toBeInTheDocument();
  });
});
