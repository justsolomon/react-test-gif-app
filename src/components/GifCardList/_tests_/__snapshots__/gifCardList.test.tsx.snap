// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GifCardList /> should render the GIF card list component properly 1`] = `
<DocumentFragment>
  <div
    class="card-list"
    data-testid="card-list"
  >
    <div
      class="card-list__column"
      style="width: 25%;"
    >
      <div
        class="lazyload-wrapper "
      >
        <span>
          <span
            class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          >
            ‌
          </span>
        </span>
      </div>
      <div
        class="lazyload-wrapper "
      >
        <span>
          <span
            class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          >
            ‌
          </span>
        </span>
      </div>
    </div>
    <div
      class="card-list__column"
      style="width: 25%;"
    >
      <div
        class="lazyload-wrapper "
      >
        <span>
          <span
            class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          >
            ‌
          </span>
        </span>
      </div>
    </div>
    <div
      class="card-list__column"
      style="width: 25%;"
    >
      <div
        class="lazyload-wrapper "
      >
        <span>
          <span
            class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          >
            ‌
          </span>
        </span>
      </div>
    </div>
    <div
      class="card-list__column"
      style="width: 25%;"
    >
      <div
        class="lazyload-wrapper "
      >
        <span>
          <span
            class="react-loading-skeleton css-1eg0sfo-skeletonStyles-Skeleton"
          >
            ‌
          </span>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
